src/miral/input_configuration.cpp

```diff
@@ -400,3 +410,41 @@ void miral::InputConfiguration::Keyboard::set_repeat_rate(int new_rate) {
 void miral::InputConfiguration::Keyboard::set_repeat_delay(int new_delay) {
     self->repeat_delay = new_delay;
 }
+
+void miral::InputConfiguration::Mouse::merge(InputConfiguration::Mouse const& other)
+{
+    self->merge(*other.self);
+}
+
+void miral::InputConfiguration::Touchpad::merge(InputConfiguration::Touchpad const& other)
+{
+    self->merge(*other.self);
+}
+
+void miral::InputConfiguration::Keyboard::merge(InputConfiguration::Keyboard const& other)
+{
+    self->merge(*other.self);
+}
+
+auto miral::InputConfiguration::Mouse::operator==(Mouse const& that) const -> bool
+{
+    return self->handedness == that.self->handedness && self->acceleration == that.self->acceleration &&
+           self->acceleration_bias == that.self->acceleration_bias && self->vscroll_speed == that.self->vscroll_speed &&
+           self->hscroll_speed == that.self->hscroll_speed;
+}
+
+auto miral::InputConfiguration::Touchpad::operator==(Touchpad const& that) const -> bool
+{
+    return self->disable_while_typing == that.self->disable_while_typing &&
+           self->disable_with_external_mouse == that.self->disable_with_external_mouse &&
+           self->acceleration == that.self->acceleration && self->acceleration_bias == that.self->acceleration_bias &&
+           self->vscroll_speed == that.self->vscroll_speed && self->hscroll_speed == that.self->hscroll_speed &&
+           self->click_mode == that.self->click_mode && self->scroll_mode == that.self->scroll_mode &&
+           self->tap_to_click == that.self->tap_to_click &&
+           self->middle_button_emulation == that.self->middle_button_emulation;
+}
+
+auto miral::InputConfiguration::Keyboard::operator==(Keyboard const& that) const -> bool
+{
+    return self->repeat_rate == that.self->repeat_rate && self->repeat_delay == that.self->repeat_delay;
+}
```

Comment:
If these were needed it would be more maintainable and easier to review letting the compiler generate them:

```suggestion
auto miral::InputConfiguration::Mouse::operator==(Mouse const& that) const -> bool = default;

auto miral::InputConfiguration::Touchpad::operator==(Touchpad const& that) const -> bool = default;

auto miral::InputConfiguration::Keyboard::operator==(Keyboard const& that) const -> bool = default;
```